aip/aog/3010.md

```diff
@@ -0,0 +1,59 @@
+---
+aip:
+  id: 3010
+  state: draft
+  created: 2020-01-14
+  updated: 2020-01-14
+  type: guidance
+  scope: aog
+js:
+  - /assets/js/graphviz/viz.js
+  - /assets/js/graphviz/lite.render.js
+  - /assets/js/aip/aip-graphviz.js
+permalink: /aog/3010
+---
+
+# Actions on Google Vertical Integration Webhook Format
+
+If an action is part of a vertical program that requires a webhook for Google
+to integrate with the action, the webhook format **must** follow this AIP.
+
+## Guidance
+
+### Custom method name
+
+Custom methods in a vertical program are defined by Google and implemented by
+third-party actions. Guidelines related to names of custom methods in
+[AIP-136](../0136.md) **should** be applied.
+
+- The name of the method **should** be a verb followed by a noun.
+- If word separation is required for the method name, `lowerCamelCase`
+  **should** be used.
+
+Examples:
+
+- `getAccountBalance`
+- `completeTransaction`
+
+### Base URL in Action Package
+
+The base URL configured in Action Package defines the common part of the actual
+execution URL of the webhook. The base URL **may** include a trailing slash,
+but **must** work if `:customMethodName` is appended directly. The endpoint
+**must** be HTTPS and **must** have a valid certificate.
+
+Examples:
+
+- `https://us-central1.cloudfunctions.net/myWebhook/` (**not**
+  `https://us-central1.cloudfunctions.net/myWebhook`)
+- `https://test.com/assistant/api`
+
+### Actual execution URL
+
+Vertical program **must** compute the actual execution URL by appending
```

Comment:
Does the vertical create the URL or does AEL do it for them.

Reply:
It's not clear who this section is targeted to. 3P developers? The vertical?

Reply:
In the case of Quick Actions, we append the method name to the base URL, so it's neither AEL nor the vertical.

AIP-136 uses this text "The HTTP URI must use a : character followed by the custom verb (:archive in the above example)" - maybe we should use similar phrasing?

Reply:
It's the "client of AEL", so usually vertical. Quick actions is a special case as it's in between vertical and AEL. I consider it as part of the vertical program, because using QA is a design/impl choice of vertical handling form-filling.

Reply:
I'm afraid we can adopt "followed by the custom verb" guideline yet.

AIP-136 is resource oriented (and the resource, say "book", is in the URI), so the resource name is omitted in the ":customMethodName" (i.e. it's ":archive" instead of ":archiveBook"). 

For AoG, since there's (at least currently) only one base URL for a GF (a vertical program), vertical/3P has to put the "resource" in the URI, i.e. ":completeTransaction" instead of simply ":complete").

Reply:
All SGTM